### DISCUSSION TITLE
Deleting Existing Volume/Disk & Moving Data to Other Disks

### DISCUSSION THREAD

**Discussion by warleysa** [ClickHouse/ClickHouse#46513] [Q&A] [answered]
Hi, 

I am in a situation where one of the disks/volumes attached to Clickhouse will be taken down for maintenance next week. I could not find any information that this has been done or is possible.

Is there a method or process of removing all data from a specific volume and moving it all to a new volume?

**Ideas:** Setting keep_free_space_bytes=[total size of the volume], then removing the volume from the storage_configuration on restart. Or maybe moving all parts to a different volume

Does anybody think this would work?

**Comment by mdonkers** [comment, marked as answer]
Yes, what you could do is add another disk to your configuration, and then move all parts from the old volume to the new one.

Just add a new disk with higher priority as described in our docs: https://clickhouse.com/docs/en/operations/storing-data/
Use the `policies` section with `volumes`, setting the new one first so new data will move there.

Then use the `MOVE PART` command to move from one disk to the other: https://clickhouse.com/docs/en/sql-reference/statements/alter/partition/#move-partitionpart
(query the `system.parts` table for active parts that you'll need to move)

In case you have part names starting with a `.`, leave out the database name and wrap the name in backticks instead of normal quotes. Then it should work.

**Reply by warleysa** [reply]
Thank you! I had been looking for a solution to achieve this everywhere since our database is vertically scaled with 4 disks on a very large instance (192 CPU/768GB Memory) and we needed to replace a disk.

**Comment by d-mankowski-synerise** [comment]
Here is a short memo how we approached this problem:

1. Add a new disk to the existing volume
2. Block writes to the old disk, that is, execute `SELECT * FROM system.disks` and remember value of `total_space` field  (of the disk which you want to remove)
3. In `storage.xml` set , for the old disk,`<keep_free_space_bytes>value_of_total_space</keep_free_space_bytes>` - ClickHouse will not write any  new data 
4. Now, we need to move all data from old disk to the new one - we used a simple script running in screen:

```bash
#!/bin/bash
set -ex

while true; do
    clickhouse-client --query "SELECT DISTINCT table, partition FROM system.parts WHERE disk_name LIKE 'old_disk' LIMIT 100" > /tmp/partitions.txt
    if [ "$(wc -l /tmp/partitions.txt | awk '{print $1}')" == "0" ]; then
        exit 0
    fi

    while read -r line; do
        read arr[{1..2}] <<< $(echo $line)
        clickhouse-client --query "ALTER TABLE ${arr[1]} MOVE PARTITION ${arr[2]} TO DISK 'new_disk'" --receive_timeout=6000
    done < /tmp/partitions.txt

done
```

On slow HDD disks we got something like 300-400GB per day and we didn't notice any degradation of cluster's performance.

We also changed volume's load balancing strategy to `least_used`: `<load_balancing>least_used</load_balancing>` - if I am not mistaken, this  would also automatically move all the data from the old disk (after setting `keep_free_space_bytes` to `total_space`), but at a much slower pace.

Tested with the latest 22.8 LTS
